### ISSUE TITLE
Potential performance regression for Images

### ISSUE THREAD

**Issue by MarcinZiabek** [QuestPDF/QuestPDF#583]
Another thing we noticed is that `UseOrignalImage()` was 5x slower pr PDF:

![snip kjfTrvd6@2x](https://github.com/QuestPDF/QuestPDF/assets/206726/491a9117-78ad-4f21-8156-62dddd8fb774)

_Originally posted by @johnkors in https://github.com/QuestPDF/QuestPDF/issues/571#issuecomment-1573240098_

**Comment by MarcinZiabek** [comment]
@johnkors I have tried to validate the performance problem:

I used [this repository](https://github.com/QuestPDF/QuestPDF-ExampleInvoice) and modified it to use an actual image as a logo. Its resolution is 900x300 pixels, and is in JPEG format. 

![image](https://github.com/QuestPDF/QuestPDF/assets/9263853/42debf26-6a85-44d8-891b-6ae187db2c6d)

I am performing 10 runs, each generating 1000 PDF documents. 

Results:
1. With version `2022.12.6`, I have got the following running times (in ms): 26998, 26721, 26762, 26751, 26734, 26750, 26730, 27143, 27826, 27788
2. With version `2023.5.2`: 31440, 31120, 31153, 31334, 32118, 31500, 31402, 31485, 31759, 32401. A slight increase is expected as the new version introduces image compression and resolution scaling, so to reduce PDF file size
3. With version `2023.5.2` but with image loaded only once and cached:  27166, 27243, 27181, 26786, 27138, 27092, 27084, 27118, 26986, 27136. These results are identical with the `2022.12.6` release.
4. With version `2023.5.2` but with the `UseOriginalImage` option:  27869, 26876, 27778, 27986, 27730, 27258, 27271, 27591, 27574, 27419.

I did not observe any significant performance regression, as on your chart. What are the memory constraints on your testing machine? How much RAM does it have? What garbage collector does it use?

**Comment by johnkors** [comment]
Would be useful if you could share how you used the image APIs and how it differs from #571, as it's not part of the repo you linked to.

Things I'd check:
- UseOriginalImage() vs not using UseOriginalImage()
- Different sizes of the logo: 2500x, 1024x, 512x, 256x
- Caching of the image, not caching the image

From my brief testing, the `UseOriginalImage()` was the culprit when it came to the _crashes_ , but it introduced a perf regression. 

I'll check if image size can reduce the regression in time spent to produce the pdf.

Setup:

- S2 App Service Plan on Windows
- 3.5 GB Memory, 
- 2 vCPUs.
- CoreCLR Version: 7.0.523.17405 .NET Version: 7.0.5 Description:
- .NET Core 7 / ASP.NET Core 7
- Pdf generated in an IHostedService (Hangfire jobs)
- Azure App Service plan uses IIS on Windows App Service plans

**Comment by MarcinZiabek** [comment]
What is the resolution of your image?

**Comment by MarcinZiabek** [comment]
@johnkors Great news! 😁 

I reproduced the issue using a specifically crafted PNG file: something of a very high resolution that compresses down to a tiny file (e.g., an empty image). I suppose that you have experienced the issue due to a significantly lower amount of RAM in the Azure App Service, especially compared to a typical development machine.

Then, I decided to try different image scaling algorithms using SkiaSharp API. One of them resolves the issues. I have published a new `2023.5.3` release that hopefully will solve performance and termination problems.

Also, I have integrated an example of how to utilize the new Image API to cache shared images across documents and therefore achieve lower memory pressure with higher performance. [Please take a look here](https://github.com/QuestPDF/QuestPDF-ExampleInvoice/commit/d56d741e9ceea3dece4b7cbec150a4dc552f2ca3).

Thank you so much for your patience and helping me resolve the problem. Would you please try the new version and confirm?

**Comment by johnkors** [comment]
Nice! This looks better after I

- Upgraded to 2023.5.3 to use the new image API, and cache the image resource in a static field
- Downscaled the image from high-res to something smaller

Now we see PDFs generated on average in ~100ms, previoysly well above 1s.

I've monitored the memory/cpu footprint between runs, and it was nowhere near it's ceiling - so I don't think that's the underlying issue (at least for us). We only see a spike in CPU. But again, it's not really that heavy of a load as compared to what you ran in your concurrent test.


![snip pxQ5DM5x](https://github.com/QuestPDF/QuestPDF/assets/206726/3753932d-a64b-4103-9740-c6d207904d15)

**Comment by johnkors** [comment]
Anyhow, I think the QuestPDFs could do with some details under the "Working with images" section detailing some of these findings and pitfalls 👍

**Comment by MarcinZiabek** [comment]
That's fantastic! I'm glad it is finally fixed. Thank you so much for your help.

Also, I made a note to improve this part of the documentation. As soon as I finish adjusting the license details (to make it more user-friendly), I will do this enhancement 😁
